src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,733 @@
+		CaptureVideoFinalise();
+		if (!audio_encoder.Init(sample_rate)) {
+			LOG_ERR("FFMPEG: Failed to init audio encoder");
+			audio_encoder.Free();
+			return;
+		}
+
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		mutex.unlock();
+		waiter.notify_all();
+	}
+
+	const size_t num_samples = num_sample_frames * SamplesPerFrame;
+	std::vector<int16_t> audio_data(sample_frames, sample_frames + num_samples);
+	audio_encoder.queue.BulkEnqueue(audio_data, num_samples);
+}
+
+void FfmpegEncoder::StopQueues()
+{
+	std::unique_lock<std::mutex> lock(mutex);
+
+	// Set these first so none of the threads start another iteration
+	// Before the encoders get re-initalised
+	audio_encoder.is_initalised = false;
+	video_encoder.is_initalised = false;
+	muxer.is_initalised         = false;
+
+	audio_encoder.queue.Stop();
+	video_encoder.queue.Stop();
+	while (audio_encoder.is_working || video_encoder.is_working) {
+		waiter.wait(lock);
+	}
+	muxer.queue.Stop();
+	while (muxer.is_working) {
+		waiter.wait(lock);
+	}
+
+	// At least the muxer queue has been measured
+	// to have stale items left in it if it is never started
+	// due to one of the encoders failing to init.
+	// Explicitly clear the queues out so we don't have to worry
+	// about stale work with incorrect video/audio specs for the next recording.
+	audio_encoder.queue.Clear();
+	video_encoder.queue.Clear();
+	muxer.queue.Clear();
+
+	muxer.Free();
+	audio_encoder.Free();
+	video_encoder.Free();
+}
+
+void FfmpegEncoder::StartQueues()
+{
+	audio_encoder.queue.Start();
+	video_encoder.queue.Start();
+	muxer.queue.Start();
+}
+
+void FfmpegEncoder::CaptureVideoFinalise()
+{
+	StopQueues();
+	StartQueues();
+}
+
+bool FfmpegMuxer::Init(FfmpegVideoEncoder& video_encoder,
+                       FfmpegAudioEncoder& audio_encoder)
+{
+	const int32_t output_file_index = get_next_capture_index(CaptureType::VideoMp4);
+	const std::string output_file_path = generate_capture_filename(CaptureType::VideoMp4,
+	                                                          output_file_index).string();
+
+	// Only one of these needs to be specified. We're using filename.
+	constexpr AVOutputFormat *oformat = nullptr;
+	constexpr char *format_name = nullptr;
+	avformat_alloc_output_context2(&format_context,
+	                               oformat,
+	                               format_name,
+	                               output_file_path.c_str());
+	if (!format_context) {
+		LOG_ERR("FFMPEG: Failed to allocate format context");
+		return false;
+	}
+	AVStream* video_stream = avformat_new_stream(format_context,
+	                                             video_encoder.codec);
+	if (!video_stream) {
+		LOG_ERR("FFMPEG: Failed to create video stream");
+		return false;
+	}
+	video_stream->time_base = video_encoder.codec_context->time_base;
+
+	if (avcodec_parameters_from_context(video_stream->codecpar,
+	                                    video_encoder.codec_context) < 0) {
```

Comment:
So negative values mean errors, and zero means success? Maybe introduce an `is_ffmpeg_success` constant or something and compare to that?

Reply:
There's not really a standard among all these functions.  Usually a negative value means error.  Unless it returns a pointer, then `nullptr` is error.  Often 0 is success but some of these return some positive value on success (like number of frames converted or something).

It's not the nicest API I've ever worked with :laughing: 

I think these are already plenty clear what they're doing.  Directly under that if statement you quoted is only 2 lines:  `LOG_ERR` describing exactly what went wrong and `return false;`